docs/spec/sources.md

```diff
@@ -0,0 +1,147 @@
+
+In practice, sources are an abstract concept that allow us to create declarative
+configurations through the usage of Custom Resource Definitions (CRDs) extending
+Kubernetes. Those CRDs are instantiated by creating an instance of the resource.
+It is up to the implementation of the source author to understand the best way
+to realize the source application. This could be as 1:1 deployments inside of
+Kubernetes per resource, as a single multi-tenant application, or even an
+off-cluster implementation; or all combinations in-between.
+
+There are some guidelines on implementing sources to allow cluster operators and
+tools to dynamically discover and understand source installations.
+
+## Source CRDs
+
+CRDs that are to be understood as a `source` MUST be labeled:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  labels:
+    duck.knative.dev/source: "true" # <-- required to be a source.
+```
+
+CRDs SHOULD be added to the `sources` category:
+
+```yaml
+spec:
+  names:
+    categories:
+      - sources
+```
+
+<!-- TODO(n3wscott,nacho): document the registry reqirements.
+
+### Event Type Registry
+
+To be included in the event type registry:
+
+```yaml
+todo
+```
+-->
+
+## Source RBAC
+
+Knative leverages am aggregated RBAC role to allow for controllers to check the
+status of source type resources.
+
+The [`source-observer` ClusterRole](../../config/200-source-observer-clusterrole.yaml) looks like:
+
+```yaml
+# Use this aggregated ClusterRole when you need read "Sources".
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: source-observer
+aggregationRule:
+  clusterRoleSelectors:
+    - matchLabels:
+        duck.knative.dev/source: "true"
+rules: [] # Rules are automatically filled in by the controller manager.
+```
+
+And new sources MUST include a ClusterRole as part of installing themselves into
+a cluster:
+
+```yaml
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: foos-source-observer
+  labels:
+    duck.knative.dev/source: "true"
+rules:
+  - apiGroups:
+      - example.com
+    resources:
+      - foos
+      - foos/status
+    verbs:
+      - get
+      - list
+      - watch
+```
+
+## Source Resource Shape
+
+The minimum definition of the Kubernetes resource shape is defined in the
+[Source](https://godoc.org/github.com/knative/pkg/apis/duck/v1beta1#Source)
+ducktype.
+
+### duck.Spec
+
+The `spec` field is expected to have the following minimum shape:
+
+```go
+type SourceSpec struct {
+    // Sink is a reference to an object that will resolve to a domain name or a
+    // URI directly to use as the sink.
+    Sink apisv1alpha1.Destination `json:"sink,omitempty"`
+
+    // CloudEventOverrides defines overrides to control the output format and
+    // modifications of the event sent to the sink.
+    // +optional
+    CloudEventOverrides *CloudEventOverrides `json:"ceOverrides,omitempty"`
```

Comment:
Is `CloudEventOverrides` too long?

Reply:
what would you rather have it be called?

Reply:
My opinion and why I made it `ceOverrides` is that it does feel too much to type `cloudEventOverrides` in the yaml and cloudevents uses `ce-` as the prefix for attributes. So It felt on brand and shorter.

Reply:
btw, is it really most of the time an _override_ or isn't more often an _addition_ to the existing CloudEvent ? For me _override_ implicitly implies erasing all the original values but that's not always the case, isn't it ?

What about `ceAdditions` (or `ceAddon`) or `ceModifications` as overriding is only a potential side effect for existing keys ? 

But as it seems that this naming has been already settles in `knative/pkg` may its too late to change (so just mentioning that 'override' typically means something different to me.

Reply:
I thought of this, but if it is just additions and there is a conflict, the original should win. In this case the override wins. So it is an override. It happens to be overriding empty extensions. 

The ceOverrides is a holder for a whole class of things we might want to adjust on the source impl side. we are _overriding_ the default settings.

Reply:
You are right, "addition" is not the proper name. But maybe _ceExtension_ like in OOP when a class extends another one, it also overrides same-named methods but also adds additional methods ?  So one could view this mechanism more like an inheritance mechanism, where the source's CE is the base and can be extended by the customer provided attributes ....

Reply:
Talking with Scott, I think `ceOverrides` is right for the following reasons:

* For the correlated-Source-and-Trigger case (e.g. base FaaS scenario), this attribute will be needed to set up the correlation. Therefore, this attribute will often be set on custom objects, so we should prefer a shorter name.
* Since these values can **overwrite** existing extension data if misused, a name like "overwrite" or "override" gives a hint of the risk to application data, akin to a warning label on a set of knives.